src/java.base/share/classes/jdk/internal/classfile/impl/AbstractDirectBuilder.java

```diff
@@ -54,7 +54,9 @@ public void setOriginal(M original) {
     }
 
     public void writeAttribute(Attribute<?> a) {
-        if (context.attributesProcessingOption().isAllowed(a.attributeMapper().attributeStability())) {
+        if (a instanceof UnboundAttribute ||
+            5 - a.attributeMapper().attributeStability().ordinal()
+                    > context.attributesProcessingOption().ordinal()) {
```

Comment:
Maybe extract the `5 - a.attributeMapper().attributeStability().ordinal() > context.attributesProcessingOption().ordinal()` check into a helper method, as the current expression is too noisy:
```java
import jdk.internal.classfile.AttributeMapper.AttributeStability;
import jdk.internal.classfile.Classfile.AttributesProcessingOption;

private static final int ATTRIBUTE_STABILITY_COUNT = AttributeStability.values().length;
public static boolean isAttributeAllowed(
	final AttributeStability stability,
	final AttributesProcessingOption processingOption
) {
	return ATTRIBUTE_STABILITY_COUNT - stability.ordinal() > processingOption.ordinal();
}
```

Also, the magic number should probably be replaced with a constant storing the value of `AttributeStability.values().length`.